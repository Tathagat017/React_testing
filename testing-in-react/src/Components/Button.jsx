import React from "react";
import styles from "./button.module.css";
const Button = ({ children, color, size, handleAdd }) => {
  return (
    <>
      <button
        data-testid="custom-button"
        className={`${styles[color]} ${styles[size]}`}
        onClick={() => handleAdd()}
      >
        {children}
      </button>
    </>
  );
};

export default Button;
