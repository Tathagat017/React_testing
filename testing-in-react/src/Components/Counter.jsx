import React, { useState } from "react";
import Button from "./Button";
const Counter = () => {
  const [count, setCount] = useState(0);

  const handleAdd = () => {
    setCount((prev) => prev + 1);
  };
  return (
    <>
      <h2 data-testid="counter">Counter:{count}</h2>
      <Button size={"large"} color={"blue"} handleAdd={handleAdd}>
        ADD
      </Button>
    </>
  );
};

export default Counter;
