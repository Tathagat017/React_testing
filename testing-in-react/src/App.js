import logo from "./logo.svg";
import "./App.css";
import Button from "./Components/Button";
import Counter from "./Components/Counter";

function App() {
  return (
    <div className="App">
      <Button size={"large"} color={"teal"}>
        Hello
      </Button>
      <Button size={"large"} color={"teal"}>
        Click
      </Button>
      <Counter />
    </div>
  );
}

export default App;
