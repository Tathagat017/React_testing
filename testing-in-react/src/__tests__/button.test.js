import { fireEvent, render, screen } from "@testing-library/react";
import Button from "../Components/Button";
import App from "./../App";
import Counter from "./../Components/Counter";
import renderer from "react-test-renderer";
// wrapped is describe
// test if button renders. --> it should render
//what is you want to check first --> button should be rendering --> it should be there in virtual DOM of testing -->nothing to do with localhost and react virtual DOM we are doing via run start

// hence render it first

describe("Testing Button component", () => {
  it("Button should render on DOM", () => {
    render(
      <Button color={"teal"} size={"large"}>
        Hello
      </Button>

      //   <App />
    );
    const button = screen.getByText("Hello");
    expect(button).toBeInTheDocument();
    screen.debug();
  });

  it(" custom Button should be present on DOM", () => {
    render(
      <Button color={"teal"} size={"large"}>
        Custom Button
      </Button>
    );
    const button = screen.getByTestId("custom-button");
    expect(button).toBeInTheDocument();
    screen.debug();
  });

  it("Button should have text Hello", () => {
    render(<App />); // renders on the screen of testing(testing virtual dom)
    const button = screen.getAllByTestId("custom-button"); //returns array
    // let text = button[0].textContent();
    // expect(text).toEqual("Hello");
    //w expect(button[0]).toHaveTextContent("Hello");
    expect(button[1]).toHaveTextContent("Click");
  });
});

describe("Test the Counter", () => {
  it("test counter", () => {
    render(<Counter />);
    const text = screen.getByTestId("counter");
    const button = screen.getByTestId("custom-button");
    screen.debug();
    expect(text).toBeInTheDocument();
    expect(button).toBeInTheDocument();
  });

  it("should have initial value", () => {
    render(<Counter />);
    const text = screen.getByTestId("counter");
    expect(text).toHaveTextContent("Counter:0");
    const addButton = screen.getByTestId("custom-button");
    fireEvent.click(addButton);
    // screen.debug();
    expect(text).toHaveTextContent("Counter:1");
  });

  it("should invoke a function ", () => {
    const mock = jest.fn();
    render(<Button handleAdd={mock}></Button>);
    const button = screen.getByTestId("custom-button");
    for (let i = 0; i < 2; i++) {
      fireEvent.click(button);
    }
    //expect(mock).toBeCalled();
    expect(mock).toBeCalledTimes(2);
  });

  it("should match the snapshot", () => {
    const tree = renderer
      .create(
        <Button size={"large"} color={"teal"}>
          Click Me
        </Button>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
