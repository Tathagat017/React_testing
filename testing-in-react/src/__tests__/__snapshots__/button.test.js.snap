// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the Counter should match the snapshot 1`] = `
<button
  className="teal large"
  data-testid="custom-button"
  onClick={[Function]}
>
  Click Me
</button>
`;
